pallets/otc-settlements/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+use frame_support::{pallet_prelude::*, PalletId};
+use frame_system::{
+	offchain::{SendTransactionTypes, SubmitTransaction},
+	pallet_prelude::{BlockNumberFor, OriginFor},
+	RawOrigin,
+};
+use hydradx_traits::router::{
+	AmmTradeWeights, AmountInAndOut, AssetPair, RouteProvider, RouteSpotPriceProvider, RouterT, Trade,
+};
+use orml_traits::MultiCurrency;
+use pallet_otc::weights::WeightInfo as OtcWeightInfo;
+pub use pallet_otc::OrderId;
+use sp_arithmetic::{
+	traits::{CheckedMul, Saturating},
+	ArithmeticError, FixedPointNumber, FixedU128,
+};
+use sp_runtime::{
+	offchain::{
+		storage::StorageValueRef,
+		storage_lock::{StorageLock, Time},
+		Duration,
+	},
+	traits::AccountIdConversion,
+	Perbill,
+};
+use sp_std::vec;
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarks;
+
+pub mod weights;
+
+use weights::WeightInfo;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+
+pub type Balance = u128;
+pub type NamedReserveIdentifier = [u8; 8];
+
+pub const PALLET_ID: PalletId = PalletId(*b"otcsettl");
+
+// value taken from https://github.com/substrate-developer-hub/recipes/blob/master/pallets/ocw-demo/src/lib.rs
+pub const UNSIGNED_TXS_PRIORITY: u64 = 100;
+
+/// Vector of `SortedOtcsStorageType`
+pub const OFFCHAIN_WORKER_DATA: &[u8] = b"hydradx/otc-settlements/data/";
+/// Last block number when we updated the `OFFCHAIN_WORKER_DATA`
+pub const OFFCHAIN_WORKER_DATA_LAST_UPDATE: &[u8] = b"hydradx/otc-settlements/data-last-update/";
+pub const SORTED_ORDERS_LOCK: &[u8] = b"hydradx/otc-settlements/lock/";
+pub const LOCK_TIMEOUT_EXPIRATION: u64 = 5_000; // 5 seconds
+pub const FILL_SEARCH_ITERATIONS: u32 = 40;
+
+pub type AssetIdOf<T> = <T as pallet_otc::Config>::AssetId;
+type SortedOtcsStorageType = (OrderId, FixedU128, FixedU128, FixedU128);
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_otc::Config + SendTransactionTypes<Call<Self>> {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Named reservable multi currency.
+		type Currency: MultiCurrency<Self::AccountId, CurrencyId = AssetIdOf<Self>, Balance = Balance>;
+
+		/// Router implementation.
+		type Router: RouteProvider<AssetIdOf<Self>>
+			+ RouterT<Self::RuntimeOrigin, AssetIdOf<Self>, Balance, Trade<AssetIdOf<Self>>, AmountInAndOut<Balance>>
+			+ RouteSpotPriceProvider<AssetIdOf<Self>>;
+
+		/// Account who receives the profit.
+		#[pallet::constant]
+		type ProfitReceiver: Get<Self::AccountId>;
+
+		/// Minimum profit in terms of percentage.
+		#[pallet::constant]
+		type MinProfitPercentage: Get<Perbill>;
+
+		/// Determines when we consider an arbitrage as closed.
+		#[pallet::constant]
+		type PricePrecision: Get<FixedU128>;
```

Comment:
is this ok with various decimals ?

Reply:
It's ok. We use this param together with `Price` type.